### PR TITLE
Forloop

### PR DESCRIPTION
**Pull request by jhauris**
Adds regular for loop:
```
{% for var in values %}
  - {{ var }}
{% endfor %}
```
Currently accepts vector type, if this is accepted it can easily be extended.
Includes unit tests.

### CODE REVIEW

**Comment by hughperkins** [comment]
Merged an earlier PR which causes conflicts. Dont suppose... do you mind taking a look please?

**Comment by flexferrum** [comment]
Yes. Sure.
Both our PRs address the same functionality and pieces of sources. But this PR supports only vectors of strings, while my one covers this case but actually more general, and supports lists of any supported value types.

**Comment by flexferrum** [comment]
From my point of view it could be a good idea to merge smoothly our PRs: take loop construction processing from this one and 'tuple' value type from mine.

**Comment by jhauris** [comment]
I will take a look this evening at your commits and modify this PR to fit it.

**Comment by jhauris** [comment]
What value do you think this PR can add? It looks like the modifications you made to the ForSection and parser handle the use case I was after, mostly. I may just open a new PR that allows adding standard C++ containers to a tuple type with normal setValue in addition to using TupleValue::create(). Am I not seeing something?

**Comment by flexferrum** [comment]
> What value do you think this PR can add?
There is one 'small' difference. You handle 'range' loop and 'collection' loop as two different type of loops. While I just make a container with desired number of integers. From this point of view your implementation is more efficient (for big ranges).

> I may just open a new PR that allows adding standard C++ containers to a tuple
As you wish. :) I'm planning to do the same PR when I end with the first release of my own version of Jinja2Cpp lib.

> I may just open a new PR that allows adding standard C++ containers to a tuple
I'm not sure if it is a good idea to do it this way. In this case you will need to duplicate every 'setValue' in Template with similar 'addValue' in TupleValue (in order to keep consistency).

**Comment by jhauris** [comment]
> There is one 'small' difference. You handle 'range' loop and 'collection' loop as two different type of loops. While I just make a container with desired number of integers. From this point of view your implementation is more efficient (for big ranges).

Okay, that is what I didn't understand. You have a good point, but that may take some time.

> > I may just open a new PR that allows adding standard C++ containers to a tuple
>
> I'm not sure if it is a good idea to do it this way. In this case you will need to duplicate every 'setValue' in Template with similar 'addValue' in TupleValue (in order to keep consistency).

I definitely want to avoid that scenario, so I will give it some thought; I have some ideas.

**Comment by flexferrum** [comment]
> I definitely want to avoid that scenario, so I will give it some thought; I have some ideas.

Actually, the usage of boost::variant could be the most efficient solution in this case:
https://github.com/flexferrum/autoprogrammer/blob/master/thirdparty/Jinja2Cpp/src/Jinja2CppLight.h
But it adds the dependency to the boost. which is inappropriate (for this implementation).

**Comment by jhauris** [comment]
I removed `VectorValue` and put `ForSection`, `ForRangeSection`, and the parser back the way I had them except using flexferrum's `TupleValue` where I had `VectorValue` before.

File: src/Jinja2CppLight.h [hunk 1]
```diff
@@ -213,8 +214,8 @@ class ForSection : public ControlSection {
         if( valueByName.find( varName ) != valueByName.end() ) {
             throw render_error("variable " + varName + " already exists in this context" );
         }
-        for( auto &v : values->values ) {
-            valueByName[varName] = v;
+        for (auto i = loopStart; i < loopEnd; ++i ){
+            valueByName[varName] = std::make_shared<IntValue>( i );
```
Comment (flexferrum) [inline src/Jinja2CppLight.h]: May be it's a good idea to make shared pointer to IntValule once (outside the loop). Something like this:
```c++
auto intVal = std::make_shared<IntValue>(0);
valueByName[varName] = intVal;
for (auto i = loopStart; i < loopEnd; ++i) {
     *intVal = IntValue(i);
```

And don't forget to clear 'varName' slot after the rendering (same as in line 257)

**Review by flexferrum — APPROVED** [review]

**Comment by hughperkins** [comment]
Seems I merged these in the wrong order :)

**Comment by jhauris** [comment]
No problem! Merged master into this branch.

**Comment by hughperkins** [comment]
Awesome. Thank you :)

**Pull request merged**